libs/error-stack/src/fmt.rs

```diff
@@ -133,6 +133,148 @@
+//!    Aᶜ
+//!    |
+//!    Bᵃ
+//!   / \
+//!  Cᵃ  Eᵃ
+//!  |   |
+//!  Dᶜ  Fᵃ
+//!     / \
+//!    Gᵃ  Iᶜ
+//!    |
+//!    Hᶜ
+//! ```
+//!
+//! During formatting we distinguish between two cases (for contexts):
+//!
+//! * Lists
+//! * Groups
+//!
+//! in this explanation lists are delimited by `[` and `]`, while groups are delimited by `(` and
+//! `)`.
+//!
+//! While formatting we view the [`Report`]s as a tree of [`Frame`]s, therefore the following
+//! explanation will use terminology associated with trees, every [`Frame`] is a node and can have
+//! `0..n` children, a node that has no children (a leaf) is guaranteed to be a [`Context`].
+//!
+//! A list is a list of nodes where each node in the list if the parent of the following element and
+//! has only a single child, the last element of a list, can have `0..n` children. In the examples
+//! above, `[6, 7, 8]` is considered a list, while `[1, 6]` is not, because while `1` is a parent of
+//! `6`, `1` has more than 1 child.
+//!
+//! A group is a list of nodes where each node shares a common immediate context parent that has
+//! more than `1` child, this means that `(2, 6)` is a group (they share `1` as an immediate context
+//! parent), while `(3, 4, 6)` is not. `(3, 4, 6)` share the same parent with more than 1 child
+//! (`1`), but `1` is not the immediate context parent of `3` and `4` (`2`) is. In the more detailed
+//! example `(Dᶜ, Hᶜ, Cᶜ)` is considered a group because they share the same *immediate* context
+//! parent `Aᶜ`, important to note is that we only refer to immediate context parents, `Fᵃ` is the
+//! immediate parent of `Iᶜ`, but is not a [`Context`], therefore to find the immediate context
+//! parent, we travel up the tree until we encounter our first [`Context`] node. Groups always
+//! contain lists, for the sake of clarity this explanation only shows the first element.
+//!
+//! Additional rules that apply:
+//! * lists are never empty
+//! * lists are nested in groups
+//! * groups are always preceded by lists
+//! * groups are ordered left to right
+//!
+//! Using the aforementioned delimiters for lists and groups the end result would be:
+//!
+//! Overview Tree: `[0, 1] ([2] ([3], [4, 5]), [6, 7, 8])`
+//! Detailed Tree: `[Aᶜ] ([Dᶜ], [Hᶜ], [Iᶜ])`
+//!
+//! Attachments are not ordered by insertion order but by depth in the tree. Each context uses the
+//! attachments that are it's parents until the next context node. If attachments are shared between
+//! multiple contexts, they are duplicated and output twice.
+//!
+//! ### Output Formatting
+//!
+//! Lists are guaranteed to be non-empty and have at least a single context. The context is the
+//! heading of the whole list, while all other contexts are intended. The last entry in that
+//! indentation is (if present) the group that follows, taking the detailed example this means that
+//! the following output would be rendered:
+//!
+//! ```text
+//! Aᶜ
+//! │
+//! ╰┬▶ Dᶜ
+//!  │  ├╴Bᵃ
+//!  │  ╰╴Cᵃ
```

Comment:
It would be good if you could describe why the attachments here are in reversed order. As far as I can see, we created an error with `D` as context. Then we attached `C` and `B` (in that order). After that, we changed the context to `A`

Reply:
clarified in preceding paragraph in [c27f8c6](https://github.com/hashintel/hash/pull/2156/commits/c27f8c6658293e871570c07e9948f9c70ed634f5)